### ISSUE TITLE
Go bindings should use the context package

### ISSUE THREAD

**Issue by jynnantonix** [domokit/mojo#773]
The [context package](https://godoc.org/golang.org/x/net/context) provides a common set of utilities for assigning a work context to a request made to a service.  This includes things like timeouts, deadlines, and cancellation.  See the [blog post](https://blog.golang.org/context) for an example service that uses the context package.

The generated Go bindings should make use of contexts in both the client stubs and the service interfaces to enable things like canceling requests, assigning deadlines, etc.  For reference, [this](https://github.com/grpc/grpc-go/blob/master/call.go#L103) is how grpc deals with contexts from generated code.

@vardhan-cr @crawshaw @MitchRudominer

**Comment by vardhan-cr** [comment]
@AlexandreZani is working on a new Go bindings generator;  he might be the best person to talk to about this.

**Comment by vardhan-cr** [comment]
I don't think this is something we want, atleast at the capacity you are proposing:
- Context isn't something that can only be contained within the Go bindings; other bindings would also need to support 'Context's that the Go bindings would send in interface calls.
- I don't think we should impose the overhead of the data that 'Context' carries over at the language bindings layer, especially when it won't always be used or necessary.  It should be optional, and probably be used as a layer above the bindings layer.

My suggestion would be to write your own Context mojom for your set of participating interfaces (or propose a canonical one for Mojo), and the next step would be that bindings owners would write some utility functions to deal with Contexts.

In general, i think Contexts really only become a necessity when your underlying transport (RPC) is  unreliable. This is why they are a core part of google3's rpc system (and grpc by extension), because it's something you have to deal with if you want a reliable service.

**Comment by jynnantonix** [comment]
> In general, i think Contexts really only become a necessity when your underlying transport (RPC) is unreliable. This is why they are a core part of google3's rpc system (and grpc by extension), because it's something you have to deal with if you want a reliable service.

I disagree with this premise.  Off the top of my head, here are two situations where context would be useful that have nothing to do with the reliability of the transport:

- Using deadlines to provide priority information so that the service can make scheduling decisions.
- Canceling a long running operation (like a video stream) in response to a user action.

**Comment by vardhan-cr** [comment]
Sorry, I meant that they are not a necessity for our IPC (i was just saying that grpc needs to have things like Context because their transport is unreliable).  There are of course many reasons you would to want to fan out deadlines and cancellation mechanisms, but they are not necessary for every interface implementer, so they should be treated as a optional/utility, and above the bindings layer.  The bindings layer /could/ offer it, but then all the languages bindings would need to as well for interoperability (which means this is not just a 'go bindings' request)

**Comment by jynnantonix** [comment]
I still don't agree with that.  Consider the following example: a service wants to write data to a file that comes from a NFS server.  The pipeline looks like:

    Client -> File -> NFS implementation -> Network Service

Since there is no standard context mechanism, both the file system and network service interfaces could end up with similar but mutually incompatible context definitions, which means the NFS implementation would need to convert one to the other and then make sure that deadlines and cancellations get propagated properly.  And I'm sure whoever writes that will get it perfectly correct the first time with no bugs whatsoever.

You could make the case that file system and network interfaces should agree on a common context definition.  But at that point you are asking every interface author to predict every other service that their service might interact with and come to a consensus with all those authors on a common context definition.  Or alternatively have one canonical context definition and force every interface to only use that one, which feels a bit strange to me because at that point it seems a lot simpler to just have it be a part of the bindings so that people don't have to worry about it.

I'm also skeptical of the argument that it adds extra overhead when it's not necessary.  If the mojom encoding is similar to the protobuf encoding, then missing fields are simply not included in the encoded message, which is pretty much the best outcome that you can get.  I understand that there is other overhead in maintaining the code and checking for the existence of the field in the message but to me that cost is more than offset by the benefits of having one standard context definition that's included by default on all service interfaces.

**Comment by vardhan-cr** [comment]
> If the mojom encoding is similar to the protobuf encoding, then missing fields are simply not included in the encoded message, which is pretty much the best outcome that you can get.
The mojom encoding is nothing like protobuf (it is a lot more like Cap'n Proto);  there is no "omitting missing fields".

We should define/provide a canonical mojo.Context struct definition that services can use. But again, this I think should be an additive layer so that uninterested parties don't have to use it / deal with the overhead. Interested languages can also provide some utilities to manage Contexts as they are propogated to other services, but I don't think it should be part of the mojom interface encoding.

And again, this isn't just a golang specific issue.

**Comment by j9brown** [comment]
We're probably going to end up adding some kind of context to the IPC
mechanism at some point, even if only to allow for debug tracing across
process boundaries (it can be useful to carry metadata).  We had to do this
for Binder years ago.  But that's not today and we don't know what such a
thing would need to look like anyhow.

Jeff.
